### 타이타닉 생존자 예측모델 개발 titanic 분석용 데이터셋

In [50]:
# 우선 titanic3.csv 를 불러온다.

In [51]:
%pwd

'D:\\hjh_directory\\workspace\\jupyter_notebook\\m3_분석라이브러리\\pandas'

In [52]:
# pandas 임포트후 dataset 폴더안에있는 타이타닉3 불러오기.
# 객체이름은 titanic_df
import pandas as pd
titanic_df = pd.read_csv('dataset/titanic3.csv')
titanic_df.head()

pclass  survived                                             name     sex  \
0       1         1                    Allen, Miss. Elisabeth Walton  female   
1       1         1                   Allison, Master. Hudson Trevor    male   
2       1         0                     Allison, Miss. Helen Loraine  female   
3       1         0             Allison, Mr. Hudson Joshua Creighton    male   
4       1         0  Allison, Mrs. Hudson J C (Bessie Waldo Daniels)  female   

     age  sibsp  parch  ticket      fare    cabin embarked boat   body  \
0  29.00      0      0   24160  211.3375       B5        S    2    NaN   
1   0.92      1      2  113781  151.5500  C22 C26        S   11    NaN   
2   2.00      1      2  113781  151.5500  C22 C26        S  NaN    NaN   
3  30.00      1      2  113781  151.5500  C22 C26        S  NaN  135.0   
4  25.00      1      2  113781  151.5500  C22 C26        S  NaN    NaN   

                         home.dest  
0                     St Louis, MO  
1  Montreal, PQ / Chesterville, ON  
2  Montreal, PQ / Chesterville, ON  
3  Montreal, PQ / Chesterville, ON  
4  Montreal, PQ / Chesterville, ON

In [53]:
# 그 다음 잘 나오는지, 그리고 대충 훑어 보기
titanic_df.info()
# 1309개의 요소가 있고 컬럼은 총 14개, 널값이 있는 컬럼이 꽤있고 int float object 이렇게 타입도 볼수 있음

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1309 entries, 0 to 1308
Data columns (total 14 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   pclass     1309 non-null   int64  
 1   survived   1309 non-null   int64  
 2   name       1309 non-null   object 
 3   sex        1309 non-null   object 
 4   age        1046 non-null   float64
 5   sibsp      1309 non-null   int64  
 6   parch      1309 non-null   int64  
 7   ticket     1309 non-null   object 
 8   fare       1308 non-null   float64
 9   cabin      295 non-null    object 
 10  embarked   1307 non-null   object 
 11  boat       486 non-null    object 
 12  body       121 non-null    float64
 13  home.dest  745 non-null    object 
dtypes: float64(3), int64(4), object(7)
memory usage: 143.3+ KB


#### 컬럼 정보

- survived : 생존여부(1: 생존, 0 : 사망)
- pclass : 승선권 클래스(1 : 1st, 2 : 2nd ,3 : 3rd) 
- name : 승객 이름
- sex : 승객 성별
- age : 승객 나이
- sibsp : 동반한 형제자매, 배우자 수
- parch : 동반한 부모, 자식 수
- ticket : 티켓의 고유 넘버
- fare 티켓의 요금
- cabin : 객실 번호
- embarked : 승선한 항구명(C : Cherbourg, Q : Queenstown, S : Southampton)
- boat
- body
- home.dest

In [54]:
# 그럼 이제, 분석데이터를 얼추 보았으면 분석을 위해 전처리를 하고 불필요한 컬럼 삭제등을 한다.

In [55]:
# value_counts 로 컬럼 뽑기 함수 # unique 로 컬럼 뽑기 함수
def column_uniq_vcnt(column_name):
    print(f"컬럼 : {column_name}, 유니크값 : {titanic_df[column_name].unique()}")
    print(f"컬럼 : {column_name}, 밸류카운트값 : {titanic_df[column_name].value_counts()}")

In [56]:
# 컬럼 age
column_uniq_vcnt('age')

컬럼 : age, 유니크값 : [29.    0.92  2.   30.   25.   48.   63.   39.   53.   71.   47.   18.
 24.   26.   80.     nan 50.   32.   36.   37.   42.   19.   35.   28.
 45.   40.   58.   22.   41.   44.   59.   60.   33.   17.   11.   14.
 49.   76.   46.   27.   64.   55.   70.   38.   51.   31.    4.   54.
 23.   43.   52.   16.   32.5  21.   15.   65.   28.5  45.5  56.   13.
 61.   34.    6.   57.   62.   67.    1.   12.   20.    0.83  8.    0.67
  7.    3.   36.5  18.5   5.   66.    9.    0.75 70.5  22.5   0.33  0.17
 40.5  10.   23.5  34.5  20.5  30.5  55.5  38.5  14.5  24.5  60.5  74.
  0.42 11.5  26.5 ]
컬럼 : age, 밸류카운트값 : age
24.00    47
22.00    43
21.00    41
30.00    40
18.00    39
         ..
0.33      1
22.50     1
70.50     1
0.67      1
26.50     1
Name: count, Length: 98, dtype: int64


In [57]:
# 컬럼 cabin
column_uniq_vcnt('cabin')

컬럼 : cabin, 유니크값 : ['B5' 'C22 C26' 'E12' 'D7' 'A36' 'C101' nan 'C62 C64' 'B35' 'A23'
 'B58 B60' 'D15' 'C6' 'D35' 'C148' 'C97' 'B49' 'C99' 'C52' 'T' 'A31' 'C7'
 'C103' 'D22' 'E33' 'A21' 'B10' 'B4' 'E40' 'B38' 'E24' 'B51 B53 B55'
 'B96 B98' 'C46' 'E31' 'E8' 'B61' 'B77' 'A9' 'C89' 'A14' 'E58' 'E49' 'E52'
 'E45' 'B22' 'B26' 'C85' 'E17' 'B71' 'B20' 'A34' 'C86' 'A16' 'A20' 'A18'
 'C54' 'C45' 'D20' 'A29' 'C95' 'E25' 'C111' 'C23 C25 C27' 'E36' 'D34'
 'D40' 'B39' 'B41' 'B102' 'C123' 'E63' 'C130' 'B86' 'C92' 'A5' 'C51' 'B42'
 'C91' 'C125' 'D10 D12' 'B82 B84' 'E50' 'D33' 'C83' 'B94' 'D49' 'D45'
 'B69' 'B11' 'E46' 'C39' 'B18' 'D11' 'C93' 'B28' 'C49' 'B52 B54 B56' 'E60'
 'C132' 'B37' 'D21' 'D19' 'C124' 'D17' 'B101' 'D28' 'D6' 'D9' 'B80' 'C106'
 'B79' 'C47' 'D30' 'C90' 'E38' 'C78' 'C30' 'C118' 'D36' 'D48' 'D47' 'C105'
 'B36' 'B30' 'D43' 'B24' 'C2' 'C65' 'B73' 'C104' 'C110' 'C50' 'B3' 'A24'
 'A32' 'A11' 'A10' 'B57 B59 B63 B66' 'C28' 'E44' 'A26' 'A6' 'A7' 'C31'
 'A19' 'B45' 'E34' 'B78' 'B50' 'C87' 'C1

In [58]:
# 컬럼 fare
column_uniq_vcnt('fare')

컬럼 : fare, 유니크값 : [211.3375 151.55    26.55    77.9583   0.      51.4792  49.5042 227.525
  69.3     78.85    30.      25.925  247.5208  76.2917  75.2417  52.5542
 221.7792  26.      91.0792 135.6333  35.5     31.     164.8667 262.375
  55.      30.5     50.4958  39.6     27.7208 134.5     26.2875  27.4458
 512.3292   5.      47.1    120.      61.175   53.1     86.5     29.7
 136.7792  52.      25.5875  83.1583  25.7     71.      71.2833  57.
  81.8583 106.425   56.9292  78.2667  31.6792  31.6833 110.8833  26.3875
  27.75   263.     133.65    49.5     79.2     38.5    211.5     59.4
  89.1042  34.6542  28.5    153.4625  63.3583  55.4417  76.7292  42.4
  83.475   93.5     42.5     51.8625  50.      57.9792  90.      30.6958
  80.      28.7125  25.9292  39.4     45.5    146.5208  82.1708  57.75
 113.275   26.2833 108.9     25.7417  61.9792  66.6     40.125   55.9
  60.      82.2667  32.3208  79.65    28.5375  33.5     34.0208  75.25
  77.2875  61.3792  35.      24.      13.      11.5    

In [59]:
# 컬럼 embarked
column_uniq_vcnt('embarked')

컬럼 : embarked, 유니크값 : ['S' 'C' nan 'Q']
컬럼 : embarked, 밸류카운트값 : embarked
S    914
C    270
Q    123
Name: count, dtype: int64


In [60]:
# 컬럼 home.dest
column_uniq_vcnt('body')

컬럼 : body, 유니크값 : [ nan 135.  22. 124. 148. 208. 172. 269.  62. 133. 275. 147. 110. 307.
  38.  80.  45. 258. 126. 292. 175. 249. 230. 122. 263. 234. 189. 166.
 207. 232.  16. 109.  96.  46. 245. 169. 174.  97.  18. 130.  17. 295.
 286. 236. 322. 297. 155. 305.  19.  75.  35. 256. 149. 283. 165. 108.
 121.  52. 209. 271.  43.  15. 101. 287.  81. 294. 293. 190.  72. 103.
  79. 259. 260. 142. 299. 171.   9. 197.  51. 187.  68.  47.  98. 188.
  69. 306. 120. 143. 156. 285.  37.  58.  70. 196. 153.  61.  53. 201.
 309. 181. 173.  89.   4. 206. 327. 119.   7.  32.  67. 284. 261. 176.
  50.   1. 255. 298. 314.  14. 131. 312. 328. 304.]
컬럼 : body, 밸류카운트값 : body
135.0    1
101.0    1
37.0     1
285.0    1
156.0    1
        ..
97.0     1
174.0    1
169.0    1
245.0    1
304.0    1
Name: count, Length: 121, dtype: int64


#### ○ null 처리

* Age : null값 263개를 갖고 있음
    * 전처리 : O
        * 위의 column 분석에 의하면 영향력이 적지 않으므로 null값을 적절히 처리한 후 활용할 것
* Cabin : null값 1014개를 갖고 있음
    * 전처리 : X(제외)
        * 사고 시각은 자정 즈음이었으나, 갑작스러운 사고로 배가 단숨에 침몰한 것이 아니므로 객실의 레벨과 생존여부는 관련이 없을 것으로 추측됨
        * 객실의 수준을 통해 얻을 수 있는 정보는 대표적으로 탑승객의 재력이며, 이는 Fare 등으로 대체할 수 있다고 여겨짐
* Fare : null값 1개 갖고 있음
    * 전처리 : O
        * null 값을 적절히 처리한 후 활용할 것. 티켓의 요금은 pclass 와 연관이 높을 것으로 봐서 적절히 값을 넣어주면 될듯
* Embarked : null값 2개를 갖고 있음 [거의 없음]
    * 전처리 : X(제외)
        * 생존자 예측모델에서 승선한 항구명은 관련이 없다고 판단. 제외
* boat, body, home.dest : 널 값이 상당히 많음
    * 전처리 : X(제외)
        * 값이 가치가 없어 보이고 널 값도 많고 생존 여부 모델에는 큰 관련이 없을 것으로 추측됨. 영향력이 적은 컬럼으로 제외.

In [61]:
# fare 가 널값의 개수가 1개여서 이거먼저해보자.
# fare 에서 널값이 있는 행
titanic_df[titanic_df.fare.isna()]

pclass  survived                name   sex   age  sibsp  parch ticket  \
1225       3         0  Storey, Mr. Thomas  male  60.5      0      0   3701   

      fare cabin embarked boat   body home.dest  
1225   NaN   NaN        S  NaN  261.0       NaN

In [62]:
# fare Null에 대체값 탐색
# fare 는 pclass 와 관련이 있을 것으로 보고 pclass가 3인 것을 좀더 탐색
pclass_3_df = titanic_df[titanic_df['pclass'] == 3]
pclass_3_df
# pclass가 3인 사람들의 fare 의 describe 로 확인하면 될듯
pclass_3_df.fare.describe()
# 최댓값이 평균값과 사분할 값보다 너무 격차가 커서
# 임의로 범위를 정한뒤 탐색후에 값을 정해야 할것으로 보임
# 총708, min0, max69, 평균13, std11, 25%7, 50%8, 75%15

count    708.000000
mean      13.302889
std       11.494358
min        0.000000
25%        7.750000
50%        8.050000
75%       15.245800
max       69.550000
Name: fare, dtype: float64

In [63]:
# min 값이 0 이라서 의문점을 갖고 탐색.
pclass_3_df[pclass_3_df.fare == 0]
# fare가 0 인사람은 총 4명이고, 그 이유를 검색해보니 4가지 경우가 있을수 있다.
# 1. 승무원인데 고객으로 간경우 무료로? 2. 무료티켓? 3. 데이터오류 4. 특별고객
# 즉, 요금이 0인 데이터를 따로 처리할 이유는 없어보임.
# 범위를 좁히기 위해 요금인 0인 사람은 pclass_3 에서 4명이라서 적은 경우이기 때문에 0을 제외 하는것이 맞다고 판단

pclass  survived                             name   sex   age  sibsp  \
896        3         0              Johnson, Mr. Alfred  male  49.0      0   
898        3         0  Johnson, Mr. William Cahoone Jr  male  19.0      0   
963        3         0              Leonard, Mr. Lionel  male  36.0      0   
1254       3         1     Tornquist, Mr. William Henry  male  25.0      0   

      parch ticket  fare cabin embarked boat  body home.dest  
896       0   LINE   0.0   NaN        S  NaN   NaN       NaN  
898       0   LINE   0.0   NaN        S  NaN   NaN       NaN  
963       0   LINE   0.0   NaN        S  NaN   NaN       NaN  
1254      0   LINE   0.0   NaN        S   15   NaN       NaN

In [64]:
# 0보다 큰값부터, 75% 값에 mean을 더한값까지를 범위로 하고 다시 탐색.
pclass_3_df[(pclass_3_df['fare'] > 0) & (pclass_3_df['fare'] <= 13.302889 + 15.245800)].fare.describe()
# mean 값은 10.481546, 분포가 25퍼가 7.몇, 50퍼가 7.몇 으로 가장 많아서
# 빈도수가 많은 50% 의 값인 7.925000(7.925) 으로 널값을 대체하기로 함(중앙값)

count    645.000000
mean      10.485510
std        4.934281
min        3.170800
25%        7.750000
50%        7.925000
75%       12.287500
max       27.900000
Name: fare, dtype: float64

In [65]:
# fare Null 값처리 : 7.925 (중앙값)으로 대체
import warnings
warnings.filterwarnings("ignore")
import pandas as pd
titanic_df.fare.fillna(7.925, inplace=True)
# fare null값 체크
titanic_df[titanic_df.fare.isna()]
#titanic_df.fare.isnull().any()

Empty DataFrame
Columns: [pclass, survived, name, sex, age, sibsp, parch, ticket, fare, cabin, embarked, boat, body, home.dest]
Index: []

## fare null값처리 완료. 이후는 age null값처리과정..

In [66]:
# 이번에는 age 를 전처리 해보자
# age null 값 있는 행 확인
titanic_df[titanic_df.age.isna()].head()

pclass  survived                                               name  \
15       1         0                                Baumann, Mr. John D   
37       1         1      Bradley, Mr. George ("George Arthur Brayton")   
40       1         0                          Brewe, Dr. Arthur Jackson   
46       1         0                              Cairns, Mr. Alexander   
59       1         1  Cassebeer, Mrs. Henry Arthur Jr (Eleanor Genev...   

       sex  age  sibsp  parch    ticket     fare cabin embarked boat  body  \
15    male  NaN      0      0  PC 17318  25.9250   NaN        S  NaN   NaN   
37    male  NaN      0      0    111427  26.5500   NaN        S    9   NaN   
40    male  NaN      0      0    112379  39.6000   NaN        C  NaN   NaN   
46    male  NaN      0      0    113798  31.0000   NaN        S  NaN   NaN   
59  female  NaN      0      0     17770  27.7208   NaN        C    5   NaN   

           home.dest  
15      New York, NY  
37   Los Angeles, CA  
40  Philadelphia, PA  
46               NaN  
59      New York, NY

In [67]:
# age 의 통계 확인
titanic_df.age.describe()
# 평균29, min0.25%21

count    1046.000000
mean       29.881138
std        14.413493
min         0.170000
25%        21.000000
50%        28.000000
75%        39.000000
max        80.000000
Name: age, dtype: float64

In [68]:
# 강사님이 말씀해주신 age 는 이름의 mr ms 이런거를 따서 그걸로 하면 될듯!
column_uniq_vcnt('name')
# 이름을 하나 보면 'Allison, Miss. Helen Loraine'
# , 와 . 으로 구분해서 스플릿으로 다 없애서 호칭만 남기게끔 하자

컬럼 : name, 유니크값 : ['Allen, Miss. Elisabeth Walton' 'Allison, Master. Hudson Trevor'
 'Allison, Miss. Helen Loraine' ... 'Zakarian, Mr. Mapriededer'
 'Zakarian, Mr. Ortin' 'Zimmerman, Mr. Leo']
컬럼 : name, 밸류카운트값 : name
Connolly, Miss. Kate             2
Kelly, Mr. James                 2
Allen, Miss. Elisabeth Walton    1
Ilmakangas, Miss. Ida Livija     1
Ilieff, Mr. Ylio                 1
                                ..
Hart, Miss. Eva Miriam           1
Harris, Mr. Walter               1
Harris, Mr. George               1
Harper, Rev. John                1
Zimmerman, Mr. Leo               1
Name: count, Length: 1307, dtype: int64


In [69]:
# age에서 null값이 있는것만 담기
age_null_df = titanic_df[titanic_df['age'].isnull()]
age_null_df.head()

pclass  survived                                               name  \
15       1         0                                Baumann, Mr. John D   
37       1         1      Bradley, Mr. George ("George Arthur Brayton")   
40       1         0                          Brewe, Dr. Arthur Jackson   
46       1         0                              Cairns, Mr. Alexander   
59       1         1  Cassebeer, Mrs. Henry Arthur Jr (Eleanor Genev...   

       sex  age  sibsp  parch    ticket     fare cabin embarked boat  body  \
15    male  NaN      0      0  PC 17318  25.9250   NaN        S  NaN   NaN   
37    male  NaN      0      0    111427  26.5500   NaN        S    9   NaN   
40    male  NaN      0      0    112379  39.6000   NaN        C  NaN   NaN   
46    male  NaN      0      0    113798  31.0000   NaN        S  NaN   NaN   
59  female  NaN      0      0     17770  27.7208   NaN        C    5   NaN   

           home.dest  
15      New York, NY  
37   Los Angeles, CA  
40  Philadelphia, PA  
46               NaN  
59      New York, NY

In [70]:
# split 사용해서 호칭만 담기
titles_only = age_null_df['name'].apply(lambda x: x.split(',')[1].split('.')[0].strip())

# 유니크한 호칭과 그 빈도수를 출력
unique_titles_only = titles_only.unique()
value_counts_only = titles_only.value_counts()

print(unique_titles_only)
print()
value_counts_only

# ['Mr' 'Dr' 'Mrs' 'Miss' 'Master' 'Ms'] 호칭이 이것인것들이 null 이니까 이것을 대체해줄 값을 찾자

['Mr' 'Dr' 'Mrs' 'Miss' 'Master' 'Ms']



name
Mr        176
Miss       50
Mrs        27
Master      8
Dr          1
Ms          1
Name: count, dtype: int64

In [71]:
# 해당 호칭 포함되는거 찾아보기..
# master 의 경우 age 가 굉장히 어림
str_contains_df = titanic_df[titanic_df['name'].str.contains('Master.')]
str_contains_df

pclass  survived                                 name   sex    age  \
1          1         1       Allison, Master. Hudson Trevor  male   0.92   
54         1         1  Carter, Master. William Thornton II  male  11.00   
94         1         1            Dodge, Master. Washington  male   4.00   
249        1         1          Ryerson, Master. John Borie  male  13.00   
273        1         1      Spedden, Master. Robert Douglas  male   6.00   
...      ...       ...                                  ...   ...    ...   
1207       3         0         Skoog, Master. Karl Thorsten  male  10.00   
1240       3         1      Thomas, Master. Assad Alexander  male   0.42   
1256       3         1       Touma, Master. Georges Youssef  male   7.00   
1262       3         0  van Billiard, Master. James William  male    NaN   
1263       3         0    van Billiard, Master. Walter John  male  11.50   

      sibsp  parch    ticket      fare            cabin embarked boat  body  \
1         1      2    113781  151.5500          C22 C26        S   11   NaN   
54        1      2    113760  120.0000          B96 B98        S    4   NaN   
94        0      2     33638   81.8583              A34        S    5   NaN   
249       2      2  PC 17608  262.3750  B57 B59 B63 B66        C    4   NaN   
273       0      2     16966  134.5000              E34        C    3   NaN   
...     ...    ...       ...       ...              ...      ...  ...   ...   
1207      3      2    347088   27.9000              NaN        S  NaN   NaN   
1240      0      1      2625    8.5167              NaN        C   16   NaN   
1256      1      1      2650   15.2458              NaN        C    C   NaN   
1262      1      1  A/5. 851   14.5000              NaN        S  NaN   NaN   
1263      1      1  A/5. 851   14.5000              NaN        S  NaN   1.0   

                            home.dest  
1     Montreal, PQ / Chesterville, ON  
54                      Bryn Mawr, PA  
94                  San Francisco, CA  
249   Haverford, PA / Cooperstown, NY  
273                   Tuxedo Park, NY  
...                               ...  
1207                              NaN  
1240                              NaN  
1256                              NaN  
1262                              NaN  
1263                              NaN  

[61 rows x 14 columns]

In [72]:
# ['Mr' 'Dr' 'Mrs' 'Miss' 'Master' 'Ms'] 에 해당하는 age를 분석해서 null에 대체
str_contains_df = titanic_df[titanic_df['name'].str.contains('Master.')]
round(str_contains_df.age.describe(), 2)

count    53.00
mean      5.48
std       4.16
min       0.33
25%       2.00
50%       4.00
75%       9.00
max      14.50
Name: age, dtype: float64

In [73]:
# ['Mr' 'Dr' 'Mrs' 'Miss' 'Master' 'Ms'] 에 해당하는 mean 값으로 df 만들기
import pandas as pd
mean_title = pd.DataFrame(columns=['Mr', 'Dr', 'Mrs', 'Miss', 'Master', 'Ms'], index=['평균나이'])
mean_title
mean_title.loc['평균나이', 'Mr'] = 33.32
mean_title.loc['평균나이', 'Dr'] = 37.27
mean_title.loc['평균나이', 'Mrs'] = 36.80
mean_title.loc['평균나이', 'Miss'] = 21.77
mean_title.loc['평균나이', 'Master'] = 5.48 # Master는 나이가 어린 남자 아이들
mean_title.loc['평균나이', 'Ms'] = 28.0
mean_title

Mr     Dr   Mrs   Miss Master    Ms
평균나이  33.32  37.27  36.8  21.77   5.48  28.0

In [74]:
mean_age = mean_title.loc['평균나이', 'Mr']
mean_age

33.32

In [75]:
def describe_age(title):
    str_contains_df = titanic_df[titanic_df['name'].str.contains(title)]
    return str_contains_df['age'].describe()
describe_age('Mr.')

count    756.000000
mean      33.322090
std       12.681912
min       11.000000
25%       24.000000
50%       30.000000
75%       41.000000
max       80.000000
Name: age, dtype: float64

In [76]:
# age의 null값 대체
titles = ['Mr', 'Dr', 'Mrs', 'Miss', 'Master', 'Ms']
for title in titles:
    mean_age = mean_title.loc['평균나이', title]
    titanic_df.loc[titanic_df['name'].str.contains(title) & titanic_df['age'].isnull(), 'age'] = mean_age

# age의 null값 check
titanic_df[titanic_df.age.isnull()]
#titanic_df.age.isnull().any()

Empty DataFrame
Columns: [pclass, survived, name, sex, age, sibsp, parch, ticket, fare, cabin, embarked, boat, body, home.dest]
Index: []

## 여기까지가 null 값 대체.. 이후 컬럼 삭제

In [77]:
titanic_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1309 entries, 0 to 1308
Data columns (total 14 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   pclass     1309 non-null   int64  
 1   survived   1309 non-null   int64  
 2   name       1309 non-null   object 
 3   sex        1309 non-null   object 
 4   age        1309 non-null   float64
 5   sibsp      1309 non-null   int64  
 6   parch      1309 non-null   int64  
 7   ticket     1309 non-null   object 
 8   fare       1309 non-null   float64
 9   cabin      295 non-null    object 
 10  embarked   1307 non-null   object 
 11  boat       486 non-null    object 
 12  body       121 non-null    float64
 13  home.dest  745 non-null    object 
dtypes: float64(3), int64(4), object(7)
memory usage: 143.3+ KB


#### 컬럼 정보

- survived : 생존여부(1: 생존, 0 : 사망)
- pclass : 승선권 클래스(1 : 1st, 2 : 2nd ,3 : 3rd) 
- name : 승객 이름
- sex : 승객 성별
- age : 승객 나이
- sibsp : 동반한 형제자매, 배우자 수
- parch : 동반한 부모, 자식 수
- ticket : 티켓의 고유 넘버
- fare 티켓의 요금
- cabin : 객실 번호
- embarked : 승선한 항구명(C : Cherbourg, Q : Queenstown, S : Southampton)
- boat
- body
- home.dest

- 여기서 필요 없는 컬럼은..
- boat, body, home.dest, cabin -> null값 많아서 대체시 왜곡가능성.. 만드려고하는 모델과 연관X 제외.
- name, ticket, embarked ->연관X 제외.

In [78]:
# 불필요한 컬럼 삭제
titanic_df = titanic_df.drop(['name', 'ticket', 'boat', 'body', 'home.dest', 'cabin', 'embarked'], axis=1)

In [79]:
titanic_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1309 entries, 0 to 1308
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   pclass    1309 non-null   int64  
 1   survived  1309 non-null   int64  
 2   sex       1309 non-null   object 
 3   age       1309 non-null   float64
 4   sibsp     1309 non-null   int64  
 5   parch     1309 non-null   int64  
 6   fare      1309 non-null   float64
dtypes: float64(2), int64(4), object(1)
memory usage: 71.7+ KB


In [80]:
#### 컬럼 sex 숫자로 변환
titanic_df['sex'] = titanic_df['sex'].map({'female': 1, 'male': 0})
titanic_df

pclass  survived  sex    age  sibsp  parch      fare
0          1         1    1  29.00      0      0  211.3375
1          1         1    0   0.92      1      2  151.5500
2          1         0    1   2.00      1      2  151.5500
3          1         0    0  30.00      1      2  151.5500
4          1         0    1  25.00      1      2  151.5500
...      ...       ...  ...    ...    ...    ...       ...
1304       3         0    1  14.50      1      0   14.4542
1305       3         0    1  21.77      1      0   14.4542
1306       3         0    0  26.50      0      0    7.2250
1307       3         0    0  27.00      0      0    7.2250
1308       3         0    0  29.00      0      0    7.8750

[1309 rows x 7 columns]

## 컬럼 삭제 완료. 이후는 파생변수..

In [81]:
#### 파생변수 만들기 과정

In [82]:
titanic_df.head()

pclass  survived  sex    age  sibsp  parch      fare
0       1         1    1  29.00      0      0  211.3375
1       1         1    0   0.92      1      2  151.5500
2       1         0    1   2.00      1      2  151.5500
3       1         0    0  30.00      1      2  151.5500
4       1         0    1  25.00      1      2  151.5500

In [83]:
column_uniq_vcnt('pclass')

컬럼 : pclass, 유니크값 : [1 2 3]
컬럼 : pclass, 밸류카운트값 : pclass
3    709
1    323
2    277
Name: count, dtype: int64


In [84]:
# 각 pclass에 따른 fare의 통계
fare_by_pclass = titanic_df.groupby('pclass')['fare'].describe()
fare_by_pclass

count       mean        std  min      25%      50%       75%       max
pclass                                                                        
1       323.0  87.508992  80.447178  0.0  30.6958  60.0000  107.6625  512.3292
2       277.0  21.179196  13.607122  0.0  13.0000  15.0458   26.0000   73.5000
3       709.0  13.295304  11.488014  0.0   7.7500   8.0500   15.2458   69.5500

In [85]:
# fare_cat 파생변수
def fare_cat(f):
    if f <= 15:
        cat = 4
    elif f <= 30:
        cat = 3
    elif f <= 100:
        cat = 2
    else:
        cat = 1
    return cat

titanic_df['fare_cat'] = titanic_df['fare'].apply(fare_cat)
titanic_df.head()

pclass  survived  sex    age  sibsp  parch      fare  fare_cat
0       1         1    1  29.00      0      0  211.3375         1
1       1         1    0   0.92      1      2  151.5500         1
2       1         0    1   2.00      1      2  151.5500         1
3       1         0    0  30.00      1      2  151.5500         1
4       1         0    1  25.00      1      2  151.5500         1

In [86]:
age_by_sex = titanic_df.groupby('sex')['age'].describe()
age_by_sex

count       mean        std   min   25%   50%   75%   max
sex                                                           
0    843.0  30.925872  12.907981  0.33  23.0  33.0  35.0  80.0
1    466.0  28.245129  13.531317  0.17  21.0  26.0  35.0  76.0

- 타이타닉에서 구조순서가 나이가 어린사람, 그다음 여성, 부자인가아닌가, 마지막에는 결국 가난하고 남자인사람이 생존률이 가장 적을것
- 즉, 여자라면 구조순위에서 남자보다 앞서니까 생존률이 높고 나이가 어릴수록 구조순위에서 나이가 많은사람보다 앞서니까 생존률이 높을것으로 추정하고 만들어보기. 

In [87]:
# 성별로 구분한 나이별 파생변수
def get_combined_cat(sex, age):
    cat = ''
    if age <= 10: 
        cat = 'child'
    elif sex == 1:  # 여성인 경우
        if age <= 20: 
            cat = 'female_teen'
        elif age <= 60:
            cat = 'female_adult'
        else: 
            cat = 'female_elder'
    else:  # 남성인 경우
        if age <= 20:
            cat = 'male_teen'
        elif age <= 60:
            cat = 'male_adult'
        else: 
            cat = 'male_elder'
    return cat

titanic_df['combined_cat'] = titanic_df.apply(lambda row: get_combined_cat(row['sex'], row['age']), axis=1)
titanic_df.head()

pclass  survived  sex    age  sibsp  parch      fare  fare_cat  \
0       1         1    1  29.00      0      0  211.3375         1   
1       1         1    0   0.92      1      2  151.5500         1   
2       1         0    1   2.00      1      2  151.5500         1   
3       1         0    0  30.00      1      2  151.5500         1   
4       1         0    1  25.00      1      2  151.5500         1   

   combined_cat  
0  female_adult  
1         child  
2         child  
3    male_adult  
4  female_adult

In [88]:
# 패밀리 변수 만들기
titanic_df['family'] = titanic_df['sibsp'] + titanic_df['parch']
titanic_df

pclass  survived  sex    age  sibsp  parch      fare  fare_cat  \
0          1         1    1  29.00      0      0  211.3375         1   
1          1         1    0   0.92      1      2  151.5500         1   
2          1         0    1   2.00      1      2  151.5500         1   
3          1         0    0  30.00      1      2  151.5500         1   
4          1         0    1  25.00      1      2  151.5500         1   
...      ...       ...  ...    ...    ...    ...       ...       ...   
1304       3         0    1  14.50      1      0   14.4542         4   
1305       3         0    1  21.77      1      0   14.4542         4   
1306       3         0    0  26.50      0      0    7.2250         4   
1307       3         0    0  27.00      0      0    7.2250         4   
1308       3         0    0  29.00      0      0    7.8750         4   

      combined_cat  family  
0     female_adult       0  
1            child       3  
2            child       3  
3       male_adult       3  
4     female_adult       3  
...            ...     ...  
1304   female_teen       1  
1305  female_adult       1  
1306    male_adult       0  
1307    male_adult       0  
1308    male_adult       0  

[1309 rows x 10 columns]

In [89]:
titanic_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1309 entries, 0 to 1308
Data columns (total 10 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   pclass        1309 non-null   int64  
 1   survived      1309 non-null   int64  
 2   sex           1309 non-null   int64  
 3   age           1309 non-null   float64
 4   sibsp         1309 non-null   int64  
 5   parch         1309 non-null   int64  
 6   fare          1309 non-null   float64
 7   fare_cat      1309 non-null   int64  
 8   combined_cat  1309 non-null   object 
 9   family        1309 non-null   int64  
dtypes: float64(2), int64(7), object(1)
memory usage: 102.4+ KB


# 파생변수 만들기완료.. 이후 인코딩작업

In [90]:
# onehot인코딩 combined_cat
onehot_combined_cat = pd.get_dummies(titanic_df['combined_cat'], prefix='combined')
titanic_df = pd.concat([titanic_df, onehot_combined_cat], axis=1)
titanic_df.head()

pclass  survived  sex    age  sibsp  parch      fare  fare_cat  \
0       1         1    1  29.00      0      0  211.3375         1   
1       1         1    0   0.92      1      2  151.5500         1   
2       1         0    1   2.00      1      2  151.5500         1   
3       1         0    0  30.00      1      2  151.5500         1   
4       1         0    1  25.00      1      2  151.5500         1   

   combined_cat  family  combined_child  combined_female_adult  \
0  female_adult       0           False                   True   
1         child       3            True                  False   
2         child       3            True                  False   
3    male_adult       3           False                  False   
4  female_adult       3           False                   True   

   combined_female_elder  combined_female_teen  combined_male_adult  \
0                  False                 False                False   
1                  False                 False                False   
2                  False                 False                False   
3                  False                 False                 True   
4                  False                 False                False   

   combined_male_elder  combined_male_teen  
0                False               False  
1                False               False  
2                False               False  
3                False               False  
4                False               False

In [91]:
titanic_df.combined_cat.value_counts()

combined_cat
male_adult      670
female_adult    350
child            94
male_teen        94
female_teen      68
male_elder       26
female_elder      7
Name: count, dtype: int64

In [92]:
titanic_df.fare_cat.value_counts()

fare_cat
4    674
3    292
2    259
1     84
Name: count, dtype: int64

In [93]:
# 불필요한 컬럼 제거
def drop_feature(df):
    df.drop(['age', 'fare', 'sex', 'sibsp', 'parch', 'combined_cat'], axis=1, inplace=True)
    return df

def transform_features(df):
    df = drop_feature(df)
    return df

titanic_df = transform_features(titanic_df)
titanic_df.head()

pclass  survived  fare_cat  family  combined_child  combined_female_adult  \
0       1         1         1       0           False                   True   
1       1         1         1       3            True                  False   
2       1         0         1       3            True                  False   
3       1         0         1       3           False                  False   
4       1         0         1       3           False                   True   

   combined_female_elder  combined_female_teen  combined_male_adult  \
0                  False                 False                False   
1                  False                 False                False   
2                  False                 False                False   
3                  False                 False                 True   
4                  False                 False                False   

   combined_male_elder  combined_male_teen  
0                False               False  
1                False               False  
2                False               False  
3                False               False  
4                False               False

# SAVEPOINT. 인코딩 완료. 데이터셋저장

In [94]:
# 분석용 데이터 셋
titanic_df.to_pickle('dataset/tdf.pkl')

In [95]:
# 불러오기
titanic_df = pd.read_pickle('dataset/tdf.pkl')
titanic_df.head()

pclass  survived  fare_cat  family  combined_child  combined_female_adult  \
0       1         1         1       0           False                   True   
1       1         1         1       3            True                  False   
2       1         0         1       3            True                  False   
3       1         0         1       3           False                  False   
4       1         0         1       3           False                   True   

   combined_female_elder  combined_female_teen  combined_male_adult  \
0                  False                 False                False   
1                  False                 False                False   
2                  False                 False                False   
3                  False                 False                 True   
4                  False                 False                False   

   combined_male_elder  combined_male_teen  
0                False               False  
1                False               False  
2                False               False  
3                False               False  
4                False               False

In [96]:
# 데이터셋을 훈련, 테스트로 구분하여 분리
from sklearn.model_selection import train_test_split

# 독립변수와 종속변수 분리
y_t_df = titanic_df['survived'] # 종속변수
X_t_df = titanic_df.drop('survived', axis=1) # 독립변수

# 학습용 데이터와 평가용 데이터를 8:2로 분리
X_train, X_test, y_train, y_test = train_test_split(X_t_df, y_t_df, test_size=0.2, random_state=11)

print(X_train.shape)
print(X_test.shape)

(1047, 10)
(262, 10)


In [ ]:
from sklearn.metrics import accuracy_score
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier

# 랜덤 포레스트 모델 학습 및 평가
rf_model = RandomForestClassifier()
rf_model.fit(X_train, y_train)
rf_pred = rf_model.predict(X_test)
accuracy_rf = accuracy_score(y_test, rf_pred).round(2)

# 로지스틱 회귀 모델 학습 및 평가
lr_model = LogisticRegression()
lr_model.fit(X_train, y_train)
lr_pred = lr_model.predict(X_test)
accuracy_lr = accuracy_score(y_test, lr_pred).round(2)

print('랜덤 포레스트 정확도: {}, 로지스틱 회귀 정확도: {}'.format(accuracy_rf, accuracy_lr))